### PR TITLE
docs: Add a tutorial for building SMAUG models.

### PR DESCRIPTION
**Pull request by yaoyuannnn**

### CODE REVIEW

File: docs/source/build_python_model.rst [hunk 1]
```diff
@@ -1,4 +1,104 @@
...
+    import smaug as sg
+    with sg.Graph(name="my_model", backend="SMV") as graph:
+      # Any operators instantiated within the context will be added to `graph`.
+
+When using the :class:`smaug.Graph` API to create a graph context, we need to
+give a name for the model and select the backend we want to use when running the
+model through the C++ runtime. A backend in SMAUG is a logical combination of
+hardware blocks that implements all the SMAUG operators. Refer to
+`C++ docs <doxygen_html/index.html>`_ for more details of a backend
+implmentation in the C++ runtime. Here, we choose the `SMV` backend that comes
```
Comment (xyzsam) [inline docs/source/build_python_model.rst:24]: It's worth mentioning the reference backend too.

Comment (xyzsam) [inline docs/source/build_python_model.rst:24]: implementation typo

File: docs/source/build_python_model.rst [hunk 2]
```diff
@@ -1,4 +1,104 @@
...
+    import smaug as sg
+
+    def generate_random_data(shape):
+      r = np.random.RandomState(1234)
+      return (r.rand(*shape) * 0.005).astype(np.float16)
+
+    with sg.Graph(name="my_model", backend="SMV") as graph:
+      input_tensor = sg.Tensor(
+          data_layout=sg.NHWC, tensor_data=generate_random_data((1, 28, 28, 1)))
+      conv_tensor = sg.Tensor(
```
Comment (xyzsam) [inline docs/source/build_python_model.rst:43]: conv_weights

File: docs/source/build_python_model.rst [hunk 3]
```diff
@@ -1,4 +1,104 @@
...
+          data_layout=sg.NHWC, tensor_data=generate_random_data((32, 3, 3, 1)))
+      fc_tensor = sg.Tensor(
```
Comment (xyzsam) [inline docs/source/build_python_model.rst:45]: fc_weights

File: docs/source/build_python_model.rst [hunk 4]
```diff
@@ -1,4 +1,104 @@
...
+      act = sg.nn.convolution(
+          act, conv_tensor, stride=[1, 1], padding="same", activation="relu")
+      # After the max pooling, shape of act: [1, 32, 14, 14].
+      act = sg.nn.max_pool(act, pool_size=[2, 2], stride=[2, 2])
+      # After the matrix multiply, shape of act: [1, 10].
+      act = sg.nn.mat_mul(act, fc_tensor)
+
+As we create the first operator for the model, we need to first prepare an input
+tensor and weight tensors that are used by the operator. In the example, we
+create an input tensor `input_tensor` using the :class:`smaug.Tensor` API. Here,
```
Comment (xyzsam) [inline docs/source/build_python_model.rst:60]: Suggest: "Tensors are represented by the :class:`smaug.Tensor` object.

File: docs/source/build_python_model.rst [hunk 5]
```diff
@@ -1,4 +1,104 @@
...
+      act = sg.input_data(input_tensor)
+      # After the convolution, shape of act: [1, 32, 28, 28].
+      act = sg.nn.convolution(
+          act, conv_tensor, stride=[1, 1], padding="same", activation="relu")
+      # After the max pooling, shape of act: [1, 32, 14, 14].
+      act = sg.nn.max_pool(act, pool_size=[2, 2], stride=[2, 2])
+      # After the matrix multiply, shape of act: [1, 10].
+      act = sg.nn.mat_mul(act, fc_tensor)
+
+As we create the first operator for the model, we need to first prepare an input
```
Comment (xyzsam) [inline docs/source/build_python_model.rst:58]: Note: if you only use backticks without a role, then Sphinx is just going to render it as an italicized word. You need to prefix any inline code with `:code:` for it to render as such.

File: docs/source/build_python_model.rst [hunk 6]
```diff
@@ -1,4 +1,104 @@
...
+tensor and weight tensors that are used by the operator. In the example, we
+create an input tensor `input_tensor` using the :class:`smaug.Tensor` API. Here,
+we specify the `data_layout` as `sg.NHWC`, which stands for a 4D tensor shape
+with the channel-major layout. We also specify the `tensor_data` parameter with
+a randomly generated NumPy array, with a shape of [1, 28, 28, 1]. However, the
+user can use the real weights extracted from a pretrained model. Likewise, we
+create two weight tensors that will be used by a convolution operator and a
+matrix multiply operator, respectively.
+
+The `sg.input_data` creates a data operator that simply creates a reference
```
Comment (xyzsam) [inline docs/source/build_python_model.rst:68]: Mention that the input data operator is required for any tensor that is not the output of another operator.

File: docs/source/build_python_model.rst [hunk 7]
```diff
@@ -1,4 +1,104 @@
...
+layout transformation or vice versa as per the expected layout format of the
+backend.
+
+After finishing adding operators to the model, we can now take a look at the
+summary of the model using the :func:`smaug.Graph.print_summary` API, which
+prints model-level information and operator-specific properties.
+
+.. code-block:: python
+
+    graph.print_summary()
```
Comment (xyzsam) [inline docs/source/build_python_model.rst:91]: Can you include the output so users know what information they can expect?

File: docs/source/build_python_model.rst [hunk 8]
```diff
@@ -1,4 +1,104 @@
...
+
+.. code-block:: python
+
+    graph.write_graph()
+
+This gives us two files named `my_model_topo.pbtxt` and `my_model_params.pb`,
+where the former stores all the model information except for the parameters,
+which are stored in the latter. This separation is helpful for us to quickly
+check things in the human readable topology file while still compressing as
+much as possible the oftentimes large paramaters.
```
Comment (xyzsam) [inline docs/source/build_python_model.rst:104]: Finally, you should link this up with the C++ side by showing how to run the model. You can leave the details to the C++ API which already explains this.

File: docs/source/build_python_model.rst (hunk 1 above)
Comment (yaoyuannnn) [reply docs/source/build_python_model.rst:24]: Right. Added reference backend too.

Comment (yaoyuannnn) [reply docs/source/build_python_model.rst:24]: Done.

File: docs/source/build_python_model.rst (hunk 2 above)
Comment (yaoyuannnn) [reply docs/source/build_python_model.rst:43]: Done.

File: docs/source/build_python_model.rst (hunk 3 above)
Comment (yaoyuannnn) [reply docs/source/build_python_model.rst:45]: Done.

File: docs/source/build_python_model.rst (hunk 4 above)
Comment (yaoyuannnn) [reply docs/source/build_python_model.rst:60]: Added this.

File: docs/source/build_python_model.rst (hunk 5 above)
Comment (yaoyuannnn) [reply docs/source/build_python_model.rst:58]: Got you. Fixed all the inline code.

File: docs/source/build_python_model.rst (hunk 6 above)
Comment (yaoyuannnn) [reply docs/source/build_python_model.rst:68]: Added this.

File: docs/source/build_python_model.rst (hunk 7 above)
Comment (yaoyuannnn) [reply docs/source/build_python_model.rst:91]: Added the output.

File: docs/source/build_python_model.rst (hunk 8 above)
Comment (yaoyuannnn) [reply docs/source/build_python_model.rst:104]: Right. Linked this up with the C++ side.

File: docs/source/build_python_model.rst [hunk 9]
```diff
@@ -40,65 +42,135 @@ example of building a simple 3-layer model.
...
+tensor :code:`input_tensor` using the API. Here, we specify the
+:code:`data_layout` as :code:`sg.NHWC`, which stands for a 4D tensor shape with
+the channel-major layout. We also specify the :code:`tensor_data` parameter
+with a randomly generated NumPy array, with a shape of :code:`[1, 28, 28, 1]`.
+However, the user can use the real weights extracted from a pretrained model.
+Likewise, we create two weight tensors that will be used by a convolution
+operator and a matrix multiply operator, respectively.
+
+A :func:`smaug.data_op` is required for any tensor that is not the output of
+another operator, which simply creates a reference tensor to its input data.
```
Comment (xyzsam) [inline docs/source/build_python_model.rst:72]: nit: "..., which simply creates a reference tensor.." -> move this earlier so it reads "a data op, which simply forwards an input tensor to its output, is required for any tensor...".

File: docs/source/build_python_model.rst [hunk 10]
```diff
@@ -21,8 +21,10 @@ give a name for the model and select the backend we want to use when running the
 model through the C++ runtime. A backend in SMAUG is a logical combination of
 hardware blocks that implements all the SMAUG operators. Refer to
 `C++ docs <doxygen_html/index.html>`_ for more details of a backend
-implmentation in the C++ runtime. Here, we choose the `SMV` backend that comes
-with SMAUG, which is modeled after the NVDLA architecture. Also, refer to
+implementation in the C++ runtime. Here, we choose the :code:`SMV` backend that
+comes with SMAUG, which is modeled after the NVDLA architecture. SMAUG also has
+another backend named :code:`Reference`, which is a implementation reference
```
Comment (xyzsam) [inline docs/source/build_python_model.rst]: "a reference implementation"

**Review by xyzsam — APPROVED** [review]

File: docs/source/build_python_model.rst (hunk 9 above)
Comment (yaoyuannnn) [reply docs/source/build_python_model.rst:72]: Done.

File: docs/source/build_python_model.rst (hunk 10 above)
Comment (yaoyuannnn) [reply docs/source/build_python_model.rst]: Done.

**Pull request merged**